src/d3d12/mod.rs

```diff
@@ -0,0 +1,637 @@
+    size: u64,
+    memory_block_index: usize,
+    memory_type_index: usize,
+    heap: *mut d3d12::ID3D12Heap,
+
+    name: Option<String>,
+    backtrace: Option<String>,
+}
+
+unsafe impl Send for Allocation {}
+unsafe impl Sync for Allocation {}
+
+impl Allocation {
+    pub fn chunk_id(&self) -> Option<std::num::NonZeroU64> {
+        self.chunk_id
+    }
+
+    /// Returns the [`d3d12::ID3D12Heap`] object that is backing this allocation.
+    /// This heap object can be shared with multiple other allocations and shouldn't be freed (or allocated from)
+    /// without this library, because that will lead to undefined behavior.
+    ///
+    /// # Safety
+    /// The result of this function can safely be used to pass into [`CreatePlacedResource()`]. It's exposed
+    /// for this reason. Keep in mind to also pass [`Self::offset()`] along to it.
+    pub unsafe fn heap(&self) -> *mut d3d12::ID3D12Heap {
+        self.heap
+    }
+
+    /// Returns the offset of the allocation on the ID3D12Heap.
+    /// When creating a placed resources, this offset needs to be supplied as well.
+    pub fn offset(&self) -> u64 {
+        self.offset
+    }
+
+    /// Returns the size of the allocation
+    pub fn size(&self) -> u64 {
+        self.size
+    }
+
+    pub fn is_null(&self) -> bool {
+        self.chunk_id.is_none()
+    }
+}
+
+impl Default for Allocation {
+    fn default() -> Self {
+        Self {
+            chunk_id: None,
+            offset: 0,
+            size: 0,
+            memory_block_index: !0,
+            memory_type_index: !0,
+            heap: std::ptr::null_mut(),
+            name: None,
+            backtrace: None,
+        }
+    }
+}
+
+struct MemoryBlock {
+    heap: std::ptr::NonNull<d3d12::ID3D12Heap>,
+    sub_allocator: Box<dyn allocator::SubAllocator>,
+}
+impl MemoryBlock {
+    fn new(
+        device: &mut d3d12::ID3D12Device,
+        size: u64,
+        heap_properties: &d3d12::D3D12_HEAP_PROPERTIES,
+        heap_category: HeapCategory,
+        dedicated: bool,
+    ) -> Result<Self> {
+        let heap = unsafe {
+            let mut desc = d3d12::D3D12_HEAP_DESC {
+                SizeInBytes: size,
+                Properties: *heap_properties,
+                Alignment: d3d12::D3D12_DEFAULT_MSAA_RESOURCE_PLACEMENT_ALIGNMENT as u64,
+                ..Default::default()
+            };
+            desc.Flags = match heap_category {
+                HeapCategory::All => d3d12::D3D12_HEAP_FLAG_NONE,
+                HeapCategory::Buffer => d3d12::D3D12_HEAP_FLAG_ALLOW_ONLY_BUFFERS,
+                HeapCategory::RtvDsvTexture => d3d12::D3D12_HEAP_FLAG_ALLOW_ONLY_RT_DS_TEXTURES,
+                HeapCategory::OtherTexture => d3d12::D3D12_HEAP_FLAG_ALLOW_ONLY_NON_RT_DS_TEXTURES,
+            };
+
+            let mut heap: *mut d3d12::ID3D12Heap = std::ptr::null_mut();
+
+            let hr =
+                device.CreateHeap(&desc, &d3d12::IID_ID3D12Heap, &mut heap as *mut _ as *mut _);
```

Comment:
In Breda we migrated to:

```suggestion
            let mut heap = std::ptr::null_mut();

            let hr =
                device.CreateHeap(&desc, &d3d12::IID_ID3D12Heap, &mut heap);
            let heap = : heap.cast::<d3d12::ID3D12Heap>();
```

Thoughts?